rllib/agents/slateq/slateq_torch_model.py

```diff
@@ -0,0 +1,211 @@
+from typing import Dict, List, Sequence, Tuple
+
+import gym
+from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import ModelConfigDict, TensorType
+
+torch, nn = try_import_torch()
+F = None
+if nn:
+    F = nn.functional
+
+
+class QValueModel(nn.Module):
+    """The Q-value model for SlateQ"""
+
+    def __init__(self, embedding_size: int, q_hiddens: Sequence[int]):
+        super().__init__()
+
+        # construct hidden layers
+        layers = []
+        ins = embedding_size
+        for n in q_hiddens:
+            layers.append(nn.Linear(ins, n))
+            layers.append(nn.LeakyReLU())
+            ins = n
+        layers.append(nn.Linear(ins, 1))
+        self.layers = nn.Sequential(*layers)
+
+    def forward(self, user: TensorType, doc: TensorType) -> TensorType:
+        """Evaluate the user-doc Q model
+
+        Args:
+            user (TensorType): User embedding of shape (batch_size,
+                embedding_size).
+            doc (TensorType): Doc embeddings of shape (batch_size, num_docs,
+                embedding_size).
+
+        Returns:
+            score (TensorType): q_values of shape (batch_size, num_docs + 1).
+        """
+        batch_size, num_docs, embedding_size = doc.shape
+        doc_flat = doc.view((batch_size * num_docs, embedding_size))
+        # No user features.
+        if user.shape[-1] == 0:
+            x = doc_flat
+        else:
+            user_repeated = user.repeat(num_docs, 1)
+            x = torch.cat([user_repeated, doc_flat], dim=1)
+        x = self.layers(x)
+        # Similar to Google's SlateQ implementation in RecSim, we force the
+        # Q-values to zeros if there are no clicks.
+        x_no_click = torch.zeros((batch_size, 1), device=x.device)
+        return torch.cat([x.view((batch_size, num_docs)), x_no_click], dim=1)
```

Comment:
I think this is described in the paper. We always compute the individual Q-values for each of the (possible) documents + 1 no-click placeholder. In case the user doesn't click on any document, we need to keep track of the value of that "empty doc", I guess.